biz.aQute.bndlib/src/aQute/bnd/cdi/ReferenceDef.java

```diff
@@ -0,0 +1,99 @@
+package aQute.bnd.cdi;
+
+import org.osgi.service.component.annotations.ReferenceCardinality;
+
+import aQute.bnd.osgi.Analyzer;
+import aQute.bnd.osgi.Annotation;
+import aQute.bnd.osgi.Clazz.FieldDef;
+import aQute.bnd.osgi.Descriptors.TypeRef;
+import aQute.bnd.signatures.ClassTypeSignature;
+import aQute.bnd.signatures.FieldSignature;
+import aQute.bnd.signatures.MethodSignature;
+import aQute.bnd.signatures.TypeArgument;
+
+public class ReferenceDef {
+
+	final String				service;
+	ReferenceCardinality		cardinality	= ReferenceCardinality.MANDATORY;
+
+	public ReferenceDef(TypeRef ref) {
+		service = ref.getFQN();
+		cardinality = ReferenceCardinality.MULTIPLE;
+	}
+
+	public ReferenceDef(Annotation reference, FieldDef member, Analyzer analyzer) {
+		Object value = reference.get("value");
+
+		TypeRef ref;
+
+		if (value != null) {
+			ref = (TypeRef) value;
+		}
+		else {
+			String signature = member.getSignature();
+
+			if (signature == null) {
+				signature = member.getDescriptor()
+					.toString();
+			}
+
+			if (signature != null) {
+				ClassTypeSignature type;
+				switch (reference.getElementType()) {
+					case PARAMETER : {
+						int index = reference.getTargetIndex();
+						MethodSignature sig = MethodSignature.of(signature);
+						type = (ClassTypeSignature) sig.parameterTypes[index];
+						break;
+					}
+					default : { // FIELD
+						FieldSignature sig = FieldSignature.of(signature);
+						type = (ClassTypeSignature) sig.type;
+					}
+				}
+
+				ref = analyzer.getTypeRef(type.binary);
+				String fqn = ref.getFQN();
+
+				// unwrap Provider
+				if (fqn.equals("javax.inject.Provider")) {
+					TypeArgument typeArgument = type.classType.typeArguments[0];
+					type = (ClassTypeSignature) typeArgument.type;
+					ref = analyzer.getTypeRef(type.binary);
+					fqn = ref.getFQN();
+				}
+
+				// unwrap Collection, List, Optional
+				if (fqn.equals("java.util.Collection") || fqn.equals("java.util.List")) {
+					TypeArgument typeArgument = type.classType.typeArguments[0];
+					type = (ClassTypeSignature) typeArgument.type;
+					ref = analyzer.getTypeRef(type.binary);
+					fqn = ref.getFQN();
+					cardinality = ReferenceCardinality.MULTIPLE;
+				} else if (fqn.equals("java.util.Optional")) {
+					TypeArgument typeArgument = type.classType.typeArguments[0];
+					type = (ClassTypeSignature) typeArgument.type;
+					ref = analyzer.getTypeRef(type.binary);
+					fqn = ref.getFQN();
+					cardinality = ReferenceCardinality.OPTIONAL;
+				}
+
+				if (fqn.equals("org.osgi.service.cdi.reference.BeanServiceObjects")
+					|| fqn.equals("org.osgi.framework.ServiceReference")) {
+
+					TypeArgument typeArgument = type.classType.typeArguments[0];
+					ref = analyzer.getTypeRef(((ClassTypeSignature) typeArgument.type).binary);
+				} else if (fqn.equals("java.util.Map$Entry")) {
+					TypeArgument typeArgument = type.classType.typeArguments[1];
```

Comment:
This is wrong since Entry is an inner class of Map. The type arguments for Entry are on `type.innerTypes[0].typeArguments`. By incorrectly passing method descriptors to `MethodSignature.of` you are incorrectly seeing the type arguments on `classType`. This is because method descriptors use `$` to separate outer and inner class names while signatures use `.`. This is also why you should always use the analyzer get*Signature methods since they call Descriptors which helpfully converts `$` to `.` 

https://github.com/bndtools/bnd/blob/ceba0425550d6bd6bfb572e5419488cc042ddfd9/biz.aQute.bndlib/src/aQute/bnd/osgi/Descriptors.java#L471-L473

to ease using of a method descriptor as signature.

So none of your code (including test cases) should ever call the `of` methods to make a signature. Always use the analyzer get*Signature methods. And remember, type arguments for inner classes, like Map.Entry, are on `innerTypes`.

https://github.com/bndtools/bnd/blob/ceba0425550d6bd6bfb572e5419488cc042ddfd9/biz.aQute.bndlib/src/aQute/bnd/component/AnnotationReader.java#L1018-L1021